build-tools/packages/build-cli/src/filter.ts

```diff
@@ -0,0 +1,219 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Context, Package } from "@fluidframework/build-tools";
+import path from "node:path";
+import { ReleaseGroup } from "./releaseGroups";
+
+export interface PackageSelectionCriteria {
+	/**
+	 * True if independent packages are selected; false otherwise.
+	 */
+	independentPackages: boolean;
+
+	/**
+	 * An array of release groups whose packages are selected.
+	 */
+	releaseGroups: ReleaseGroup[];
+
+	/**
+	 * An array of release groups whose root packages are selected.
+	 */
+	releaseGroupRoots: ReleaseGroup[];
+
+	/**
+	 * If set, only selects the single package in this directory.
+	 */
+	directory?: string;
+}
+
+export interface PackageFilterOptions {
+	/**
+	 * If set, filters IN packages whose scope matches the strings provided.
+	 */
+	scope?: string[];
+	/**
+	 * If set, filters OUT packages whose scope matches the strings provided.
+	 */
+	skipScope?: string[];
+
+	/**
+	 * If set, filters private packages in/out.
+	 */
+	private: boolean | undefined;
+}
+
+/**
+ * Parses {@link selectionFlags} into a typed object that is more ergonomic than working with the flag values directly.
+ *
+ * @param flags - The parsed command flags.
+ */
+export const parsePackageSelectionFlags = (flags: any): PackageSelectionCriteria => {
+	const options: PackageSelectionCriteria =
+		flags.all === true
+			? {
+					independentPackages: true,
+					releaseGroups: ["all"],
+					releaseGroupRoots: ["all"],
+					directory: undefined,
+			  }
+			: {
+					independentPackages: flags.packages ?? false,
+					releaseGroups: flags.releaseGroup ?? [],
+					releaseGroupRoots: flags.releaseGroupRoot ?? [],
+					directory: flags.directory,
+			  };
+
+	return options;
+};
+
+/**
+ * Parses {@link filterFlags} into a typed object that is more ergonomic than working with the flag values directly.
+ *
+ * @param flags - The parsed command flags.
+ */
+export const parsePackageFilterFlags = (flags: any): PackageFilterOptions => {
+	const options: PackageFilterOptions = {
+		private: flags.private,
+		scope: flags.scope,
+		skipScope: flags.skipScope,
+	};
+
+	return options;
+};
+
+/**
+ * Selects packages from the context based on the selection. The selected packages will be filtered by the filter
+ * criteria if provided.
+ *
+ * @param context - The context.
+ * @param selection - The selection criteria to use to select packages.
+ * @param filter - An optional filter criteria to filter selected packages by.
+ * @returns An object containing the selected packages and the filtered packages.
+ */
+export const selectAndFilterPackages = (
```

Comment:
I think I'd try to `return` in between selection kinds if we know they're supposed to be mutually exclusive. We get some protection from the exclusiveness of flags, but I feel like it'd be worth it to make it explicit here to avoid bugs. Won't block the PR on it, though.

Reply:
I thought about this, but I didn't implement it intentionally. This logic is intentionally serparate from the flags. This function is processing the selection/filter criteria, which MAY come from flags, but could be from some other source. You can absolutely create a PackageSelectionCriteria that selects everything -- and that's what the --all flag does.